### PR TITLE
Tracked based Service on matches

### PR DESCRIPTION
**Pull request by snewcomer**
- Remove `this.get`
- Swap EmberArray with native array
- Reset matches array on update instead of mutating original (otherwise we need [`tracked-built-ins`](https://github.com/pzuraq/tracked-built-ins))
- `@tracked` on `matches` so that any consuming app `getter` will re-run when updated
  - https://guides.emberjs.com/release/upgrading/current-edition/tracked-properties/
  - This is a `pull` based system vs `push` based

### CODE REVIEW

File: addon/services/media.js [hunk 1]
```diff
@@ -77,35 +77,27 @@ import Evented from '@ember/object/evented';
 * @class     Media
 * @extends   Ember.Object
 */
-export default Service.extend(Evented, {
-  // eslint-disable-next-line ember/no-ember-testing-in-module-scope
-  _mocked: Ember.testing,
-  _mockedBreakpoint: 'desktop',
```
Comment (snewcomer) [inline addon/services/media.js:83]: https://github.com/ember-cli/eslint-plugin-ember/pull/272/files/36b68e3e8d9ffef76ff1bcfac9cd8154af316ee4

> Since ember-cli-qunit@4.1.0 / ember-qunit@3.0.0, `Ember.testing` is only set to
true while a test is executing instead of all the time.

Comment (snewcomer) [reply addon/services/media.js:83]: Previous addition 

https://github.com/freshbooks/ember-responsive/pull/124

> During tests, the media service will be in "desktop" unless told otherwise.

Basically this is only true initially then Ember.testing is set to `false`

File: addon/services/media.js [hunk 2]
```diff
@@ -124,21 +120,30 @@ export default Service.extend(Evented, {
...
     if (breakpoints) {
       Object.keys(breakpoints).forEach((name) => {
         const cpName = `is${classify(name)}`;
-        defineProperty(this, cpName, computed('matches.[]', function () {
-          return this.get('matches').indexOf(name) > -1;
-        }));
-        defineProperty(this, name, computed(cpName, function () {
-          return this.get(cpName);
-        }));
+        defineProperty(this, cpName, {
```
[the remark this answers is not in the document]
Comment (snewcomer) [reply addon/services/media.js:132]: native getter.  Since `matches` is tracked, this `getter` will (presumably) be asked for again when `matches` is set to a new value.

File: addon/helpers/media.js [hunk 3]
```diff
@@ -5,7 +5,7 @@ import { get } from '@ember/object';
 export default Helper.extend({
```
[the remark this answers is not in the document]
Comment (snewcomer) [reply addon/helpers/media.js]: done!

File: addon-test-support/index.js [hunk 4]
```diff
@@ -18,10 +18,9 @@ export function setBreakpoint(breakpoint) {
     }
   }
 
-  let matches = media.get('matches');
   run(() => {
-    matches.clear();
-    matches.addObjects(breakpointArray);
+    media.matches = [];
+    media.matches = breakpointArray;
```
[the remark this answers is not in the document]
Comment (snewcomer) [reply addon-test-support/index.js:22]: Probably don't need to do this anymore.

File: addon/services/media.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (snewcomer) [reply addon/services/media.js:132]: for future readers - computed properties somewhat at odd with `@tracked` and a pull based system.

File: tests/unit/services/media-test.js [hunk 5]
```diff
@@ -16,64 +15,64 @@ module('Unit | Service | media', function(hooks) {
...
-      subject.match('mobile', 'all');
-      subject.match('all', 'all');
-      subject.match('none', 'not all');
-    });
+    subject.match('mobile', 'all');
+    subject.match('all', 'all');
+    subject.match('none', 'not all');
 
-    assert.deepEqual(subject.get('matches').toArray(), ['mobile', 'all']);
+    assert.deepEqual(subject.get('matches'), ['desktop', 'mobile', 'all']);
```
Comment (alexBaizeau) [inline tests/unit/services/media-test.js:39]: Why did you need to add `desktop` here?

Comment (snewcomer) [reply tests/unit/services/media-test.js:39]: Still looking into this. Basically since/after 2017 (when the test helper was added), the instance of the service (therefore `init`) will be stamped out before `beforeEach` is run.  Looking into a solution...

File: addon-test-support/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbroeren) [reply addon-test-support/index.js:22]: This seems to be resolved now

File: addon/services/media.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (snewcomer) [reply addon/services/media.js]: Changed to `Object.defineProperty`!!

File: tests/unit/services/media-test.js (hunk 5 above)
Comment (snewcomer) [reply tests/unit/services/media-test.js]: @alexBaizeau This is fixed. 

https://github.com/freshbooks/ember-responsive/pull/286/commits/a5b5334167ae63885af7381cff6c16e7ca6537f6

File: addon/services/media.js [hunk 6]
```diff
@@ -183,27 +203,30 @@ export default Service.extend(Evented, {
...
+      if (this.isDestroyed) {
         return;
       }
 
-      this.set(`matchers.${name}`, matcher);
+      set(this, `matchers.${name}`, matcher);
 
       if (matcher.matches) {
-        this.get('matches').addObject(name);
+        this.matches = Array.from(new Set([...this.matches, name]));
```
Comment (cbroeren) [inline addon/services/media.js:222]: Is this change to trigger native getters on the matches array change?

File: tests/unit/services/media-test.js [hunk 7]
```diff
@@ -16,64 +15,64 @@ module('Unit | Service | media', function(hooks) {
...
-      subject.match('mobile', 'all');
-      subject.match('all', 'all');
-      subject.match('none', 'not all');
-    });
+    subject.match('mobile', 'all');
+    subject.match('all', 'all');
+    subject.match('none', 'not all');
 
-    assert.deepEqual(subject.get('matches').toArray(), ['mobile', 'all']);
+    assert.deepEqual(subject.get('matches'), ['mobile', 'all']);
```
Comment (cbroeren) [inline tests/unit/services/media-test.js:39]: Do you still need to use `subject.get('matches')` or could you use `subject.matches` now?

File: addon/services/media.js [hunk 8]
```diff
@@ -77,35 +77,48 @@ import Evented from '@ember/object/evented';
...
+    if (this._matches) {
+      return this._matches
+    }
+
+    return (Ember.testing && this._mocked) ? [this._mockedBreakpoint] : [];
+  }
+
+  set matches(value) {
+    this._matches = value;
+  }
```
Comment (snewcomer) [inline addon/services/media.js:108]: These getters/setters + `@tracked _matches`.

First, `matches` can't be set in init and needs to be asked for lazily.  This is b/c `setBreakpoints` is run after init.  If a user has `setBreakpoints('auto')`, then it will still have the default testing `desktop` value.

Second, `Ember.testing && this._mocked` is necessary due to the change the you should never assign Ember.testing to a variable anymore.  `this._mocked` is used to turn off the condition in testing if `setBreakpoints('auto')` is set by the user.

File: tests/unit/services/media-test.js (hunk 7 above)
Comment (snewcomer) [reply tests/unit/services/media-test.js]: Great idea!

File: addon/services/media.js (hunk 6 above)
Comment (snewcomer) [reply addon/services/media.js:222]: Yep!  reset the value instead of pushing into it.  Otherwise we need tracked-built-ins.  Since this data structure is very low footprint, this seems reasonable.

**Comment by alexBaizeau** [comment]
Thank you for your work! 👏

**Comment by snewcomer** [comment]
@alexBaizeau Thanks a lot.  We have confirmed this tracked based method works!!  (e.g. remove `computed('media.matches.[]'` and the logic updates on resize.  Looking forward to the major version bump!
